File: _kbtest.js
const { chromium } = require('/tmp/pw001/node_modules/playwright');
const { spawn } = require('child_process');
const net = require('net');

function waitPort(port) {
  return new Promise((res) => {
    const t = setInterval(() => {
      const s = net.connect(port, '127.0.0.1');
      s.on('connect', () => { clearInterval(t); s.destroy(); res(); });
      s.on('error', () => s.destroy());
    }, 100);
  });
}

(async () => {
  const srv = spawn('bash', ['-c', 'node - < server.js'], { stdio: 'ignore' });
  await waitPort(3000);
  const b = await chromium.launch();
  const pg = await b.newPage();
  const errs = [];
  pg.on('console', (m) => { if (m.type() === 'error') errs.push(m.text()); });
  await pg.goto('http://127.0.0.1:3000/', { waitUntil: 'networkidle' });

  // REQ1: empty field focused -> j/k/1-4 type as chars
  await pg.click('#intake-answer');
  await pg.fill('#intake-answer', '');
  for (const ch of ['j', 'k', '1', '2', '3', '4']) await pg.keyboard.press(ch);
  const typed = await pg.$eval('#intake-answer', (el) => el.value);
  console.log('REQ1 typed =', JSON.stringify(typed));
  if (typed !== 'jk1234') throw new Error('REQ1 FAIL: got ' + JSON.stringify(typed));

  // REQ2: no field focus -> depth-jump navigation. Fresh reload, start at top.
  await pg.reload({ waitUntil: 'networkidle' });
  await pg.evaluate(() => window.scrollTo(0, 0));
  await pg.waitForTimeout(300);
  // confirm no field is focused
  const active = await pg.evaluate(() => document.activeElement && document.activeElement.tagName);
  console.log('REQ2 activeElement tag =', active);
  const y0 = await pg.evaluate(() => window.scrollY);
  await pg.keyboard.press('j');
  await pg.waitForTimeout(700);
  const y1 = await pg.evaluate(() => window.scrollY);
  console.log('REQ2 scrollY', y0, '->', y1);
  if (!(y1 > y0)) throw new Error('REQ2 FAIL: j did not navigate down ' + y0 + '->' + y1);

  // REQ2b: numeric jump '1' back to depth 1 (top), '4' to a deeper depth
  await pg.keyboard.press('1'); await pg.waitForTimeout(700);
  const yA = await pg.evaluate(() => window.scrollY);
  await pg.keyboard.press('4'); await pg.waitForTimeout(700);
  const yB = await pg.evaluate(() => window.scrollY);
  console.log('REQ2b 1->', yA, ' 4->', yB);
  if (yB <= yA) throw new Error('REQ2b FAIL: numeric jump not advancing 1->' + yA + ' 4->' + yB);

  console.log('console errors:', errs);
  console.log('ALL PASS');
  await b.close();
  srv.kill();
  process.exit(0);
})().catch((e) => { console.error('FAILED:', e.message); process.exit(1); });

File: _kbtest.py
from playwright.sync_api import sync_playwright
import subprocess, time, socket

srv = subprocess.Popen("node - < server.js", shell=True)
for _ in range(30):
    try:
        socket.create_connection(("127.0.0.1", 3000), 0.2).close(); break
    except OSError:
        time.sleep(0.1)

def run():
  with sync_playwright() as p:
    b = p.chromium.launch()
    pg = b.new_page()
    errs = []
    pg.on("console", lambda m: errs.append(m.text) if m.type == "error" else None)
    pg.goto("http://127.0.0.1:3000/", wait_until="networkidle")
    ta = pg.query_selector("#intake-answer")
    assert ta, "no #intake-answer textarea"

    # REQ1: empty field focused -> j/k/1-4 type as chars
    ta.click(); ta.fill("")
    for ch in ["j", "k", "1", "2", "3", "4"]:
        pg.keyboard.press(ch)
    typed = pg.eval_on_selector("#intake-answer", "el=>el.value")
    print("REQ1 typed value =", repr(typed))
    assert typed == "jk1234", f"REQ1 FAIL: expected 'jk1234' got {typed!r}"

    # REQ2: focus NOT in field -> keys do depth-jump (scroll changes)
    pg.eval_on_selector("#intake-answer", "el=>el.blur()")
    pg.evaluate("window.scrollTo(0,0)")
    y0 = pg.evaluate("window.scrollY")
    pg.keyboard.press("j")
    time.sleep(0.6)
    y1 = pg.evaluate("window.scrollY")
    print(f"REQ2 scrollY {y0} -> {y1}")
    assert y1 > y0, f"REQ2 FAIL: 'j' did not navigate (scroll {y0}->{y1})"

    # REQ2b: numeric jump
    pg.keyboard.press("1"); time.sleep(0.4); yA = pg.evaluate("window.scrollY")
    pg.keyboard.press("4"); time.sleep(0.6); yB = pg.evaluate("window.scrollY")
    print(f"REQ2b '1'->{yA}  '4'->{yB}")
    assert yB != yA, "REQ2b FAIL: numeric depth-jump inert"

    print("console errors:", errs)
    print("ALL PASS")
    b.close()

run()
srv.terminate()
